DOC Merge plot_svm_margin.py and plot_separating_hyperplane.py into plot_svm_hyperplane_margin.py

SwathiR1999 opened the pull request:
### Reference Issues/PRs   
References **#30621**

### What does this implement/fix? Explain your changes.  
- Consolidated two SVM example scripts (`plot_svm_margin.py` and `plot_separating_hyperplane.py`) into a single file `plot_svm_hyperplane_margin.py`.  
- Updated explanation for clarity.  

### Any other comments?  
This PR addresses the feedback from **#30975**, which was closed. 

@StefanieSenger Looking forward to your feedback. Thanks!

SwathiR1999 committed:
Merged plot_svm_margin.py and plot_separating_hyperplane.py into a single example

SwathiR1999 committed:
Merge branch 'scikit-learn:main' into doc_addlink

SwathiR1999 committed:
Formatted code using Black and fixed linting issues with Ruff

SwathiR1999 committed:
Merge branch 'main' into doc_addlink

StefanieSenger commented on examples/svm/plot_svm_hyperplane_margin.py:
```diff
@@ -0,0 +1,69 @@
+in a two-class separable dataset using a Support Vector Machine (SVM)
+with a linear kernel and how different values of `C` influence the margin width.
+
+- **Small C (e.g., 0.05)**:
+    - Allows some misclassifications, resulting in a wider margin.
+- **Moderate C (e.g., 1)**:
+    - Balances classification accuracy and margin width.
+- **Large C (e.g., 1000)**:
+    - Prioritizes classifying all points correctly, leading to a narrower margin.
```
I think while reworking the example, we could add some more information.

Do you think you could add some suggestion of when to use a small, a moderate or a high C value in a practical use case? 

We could also talk about overfitting and underfitting in this context.

Also, we could talk about miss-classification / error for different margins.

I think this could be added after the code for the plot using the `# %` syntax to make this example more notebook-like. (Also have a look at #30028 for a good example on the notebook style.)

StefanieSenger commented on examples/svm/plot_svm_hyperplane_margin.py:
```diff
@@ -0,0 +1,69 @@
+"""
+=========================================================================
+SVM: Effect of Regularization (C) on Maximum Margin Separating Hyperplane
+========================================================================
```
```suggestion
=========================================================================
```

The lacking "=" leads to the title not being rendered correctly. Adding "=" should fix it.

I also hope that this fixes the problem with the example not being clickable and the text overflow.

StefanieSenger commented on examples/svm/plot_svm_hyperplane_margin.py:
```diff
@@ -0,0 +1,69 @@
+        s=100,
+        linewidth=1.5,
+        facecolors="none",
+        edgecolors="r",
```
What do you think about having a different color for miss-classified samples?

StefanieSenger commented on examples/svm/plot_svm_hyperplane_margin.py:
```diff
@@ -0,0 +1,69 @@
+    clf.fit(X, y)
+
+    plt.subplot(1, 3, i)
+    plt.scatter(X[:, 0], X[:, 1], c=y, s=30, cmap=plt.cm.Paired, edgecolors="k")
```
Please add axis labels. Something simple like "Feature 1" and "Feature 2" would be enough.

StefanieSenger replied:
This is resolved. (Just the "resolved" button doesn't show here.)

StefanieSenger commented on examples/svm/plot_svm_hyperplane_margin.py:
```diff
@@ -0,0 +1,69 @@
+        s=100,
+        linewidth=1.5,
+        facecolors="none",
+        edgecolors="r",
```
What do you think about having a different color for miss-classified samples?

StefanieSenger replied:
This is resolved. (Just the "resolved" button doesn't show here.)

StefanieSenger reviewed (commented):
Hi @SwathiR1999,

thank you for your patience and for your efforts to merge the examples. It looks very good and I like the design of the plots. The plots all have heading and legend and are easy on the eye.

I think we could add some information and I have left you some suggestions on that, too.

When I build the docs locally, I cannot click to open it and also there is some text overflow in the background. I believe this is because of a rending issue in the heading, and I have added a suggestions for a fix.

What is lacking is to remove the old examples and to add redirections from both old links to the new example link in the `doc/conf.py` file. You can have a look at #30028 to see how to add the redirections.

You need to delete all links to the old examples as well (and replace it with the new one if suitable). There are some, probably too many, links to `plot_separating_hyperplane_unbalanced` currently.

SwathiR1999 committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into doc_addlink

SwathiR1999 committed:
Merge branch 'doc_addlink' of https://github.com/SwathiR1999/scikit-learn into doc_addlink

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated conf.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

SwathiR1999 commented:
Hi @StefanieSenger  , thanks for the feedback!

> There are some, probably too many, links to `plot_separating_hyperplane_unbalanced` currently.

Just to clarify - should I also apply similar changes (merge/update/remove references) to `plot_separating_hyperplane_unbalanced.py`, or is the current focus only on replacing `plot_svm_margin.py` and `plot_separating_hyperplane.py` with the new `plot_svm_hyperplane_margin.py` example?

SwathiR1999 committed:
updated plot_svm_hyperplane_margin.py

StefanieSenger commented:
> Just to clarify - should I also apply similar changes (merge/update/remove references) to plot_separating_hyperplane_unbalanced.py, or is the current focus only on replacing plot_svm_margin.py and plot_separating_hyperplane.py with the new plot_svm_hyperplane_margin.py example?

Oh, my bad. I didn't see there were `plot_separating_hyperplane_unbalanced.py` AND `plot_separating_hyperplane.py`.
No please keep the current focus and don't let you be confused by me. :)

SwathiR1999 committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into doc_addlink

SwathiR1999 committed:
formatted plot_svm_hyperplane_margin.py

SwathiR1999 committed:
formatted plot_svm_hyperplane_margin.py

SwathiR1999 committed:
formatted plot_svm_hyperplane_margin.py

SwathiR1999 committed:
Update SVM docs and remove deprecated example

- Modified \`doc/conf.py\` and \`doc/modules/svm.rst\` to update documentation and add example reference.\n- Removed outdated example: \`examples/svm/plot_svm_margin.py\`.

SwathiR1999 committed:
modified svm.rst

SwathiR1999 committed:
modified svm.rst

SwathiR1999 commented:
Hi @StefanieSenger,

I’ve implemented your suggestions. I wanted to check with you on one small point:

While `plot_svm_margin.py` is fully covered by the merged version and can be removed, I feel that `plot_separating_hyperplane.py` still serves a valuable purpose. It provides a clean and minimal visual introduction to the concept of an SVM decision boundary and support vectors, which may be helpful for first-time readers.

It is currently referenced early in the SVM documentation, which aligns well with its simplicity. The new merged example, on the other hand, focuses more on the effect of `C`, making it a good fit for a later section - ideally where `C` is discussed in more detail.

Would it be alright to:

1. Retain `plot_separating_hyperplane.py` as an introductory visual example?
2. If so, would you suggest improving clarity by adding axis labels (like `'Feature 1'`, `'Feature 2'`) and annotations such as `'Support Vectors'`?
3. Add a cross-reference to the new example (`plot_svm_hyperplane_margin.py`) later in the documentation, once the role of `C` has been introduced?

Looking forward to your thoughts!

StefanieSenger reviewed (commented):
Hi @SwathiR1999,

thank you for your further work. I went through the example and left the next round of comments. (I might have more afterwards. :grimacing:)

In general, please feel very free to add any information you think would help people who start learning on SVCs and regularization to understand. We're very happy to add new passages if they are helpful and fitting the context of the example.

On your question:

> Retain plot_separating_hyperplane.py as an introductory visual example?

I can definitely see what you mean and sort of agree. But on the other hand, we have too many examples to take care of and plot_separating_hyperplane.py is pretty similar to the re-newed example you are working on. This new example would (with your help) ideally grow into a demonstration that also helps beginners. 

Maybe @adrinjalali has a different opinion on it?

StefanieSenger commented:
Hello @SwathiR1999, gentle ping to know if are you continuing to work on this PR?
Best!

SwathiR1999 commented:
Hi @StefanieSenger  .. Apologies for the delay. I had to step away for a bit due to a medical reason. Will get back to this and update the PR shortly.

StefanieSenger commented:
No worries, take your time, @SwathiR1999! I'm happy to support when needed.

SwathiR1999 committed:
Update doc/modules/svm.rst

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

SwathiR1999 committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn into doc_addlink

SwathiR1999 committed:
Update doc/modules/svm.rst

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

SwathiR1999 committed:
Update doc/modules/svm.rst

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

SwathiR1999 committed:
Merge branch 'doc_addlink' of https://github.com/SwathiR1999/scikit-learn into doc_addlink

SwathiR1999 committed:
added the reference to the example in the text.

SwathiR1999 committed:
finished the last sentence with a colon

SwathiR1999 committed:
added target line

Co-authored-by: Stefanie Senger <91849487+StefanieSenger@users.noreply.github.com>

SwathiR1999 committed:
Merge branches 'doc_addlink' and 'doc_addlink' of https://github.com/SwathiR1999/scikit-learn into doc_addlink

SwathiR1999 committed:
modified plot_svm_hyperplane_margin.py
